### PR TITLE
Architecture documentation

### PR DESCRIPTION
**Pull request by rbren**

### CODE REVIEW

File: opendevin/README.md [hunk 1]
```diff
@@ -1,25 +1,50 @@
...
+This directory contains the core components of OpenDevin.
+
+## Classes
+The key classes in OpenDevin are:
+* LLM: brokers all interactions with large language models. Works with any underlying completion model, thanks to LiteLLM.
+* Agent: responsible for looking at the current State, and producing an Action that moves one step closer toward the end-goal.
+* AgentController: initializes the Agent, manages State, and drive the main loop that pushes the Agent forward, step by step
+* State: represents the current state of the Agent's task. Includes things like the current step, a history of recent events, the Agent's long-term plan, etc
+* EventStream: a central hub for Events, where any component can publish Events, or listen for Events published by other components
+  * Event: an Action or Observeration
```
Comment (enyst) [inline opendevin/README.md:12]:
```suggestion
  * Event: an Action or Observation
```

**Review by xingyaoww — APPROVED** [review]
This LGTM! Maybe we can get this merged.

File: opendevin/README.md [hunk 2]
```diff
@@ -18,15 +19,32 @@ The key classes in OpenDevin are:
...
   AgentController--Actions-->EventStream
   EventStream--Observations-->AgentController
   Runtime--Observations-->EventStream
   EventStream--Actions-->Runtime
   Frontend--Actions-->EventStream
 ```
+
+## Runtime
+The Runtime class is abstract, and has a few different implementations:
+* We have a LocalRuntime, which runs commands and edits files directly on the user's machine
```
Comment (tobitege) [inline opendevin/README.md:48]: Nit: we should lint markdown. I'd add a blank line above each list.

File: opendevin/README.md [hunk 3]
```diff
@@ -18,15 +19,32 @@ The key classes in OpenDevin are:
...
+* We have a DockerRuntime, which runs commands inside of a docker sandbox, and edits files directly on the user's machine
+* We have an E2BRuntime, which uses e2b.dev containers to sandbox file and command operations
```
Comment (tobitege) [inline opendevin/README.md:50]: What is "E2B" to the uninitiated?

**Review by tobitege — APPROVED** [review]
LGTM, just nits below.

File: opendevin/README.md (hunk 3 above)
Comment (enyst) [reply opendevin/README.md:50]: https://github.com/e2b-dev/e2b

Maybe we can add a link somewhere 😅

File: opendevin/README.md [hunk 4]
```diff
@@ -1,25 +1,50 @@
...
-It will map `./workspace` into the docker container with the folder permission correctly adjusted for current user.
-
-Example screenshot:
-
-<img width="868" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/38853559/8dedcdee-437a-4469-870f-be29ca2b7c32">
+## Runtime
+The Runtime class is abstract, and has a few different implementations:
+* We have a LocalRuntime, which runs commands and edits files directly on the user's machine
+* We have a DockerRuntime, which runs commands inside of a docker sandbox, and edits files directly on the user's machine
+* We have an E2BRuntime, which uses e2b.dev containers to sandbox file and command operations
```
Comment (rbren) [inline opendevin/README.md:50]:
```suggestion
* We have an E2BRuntime, which uses [e2b.dev containers](https://github.com/e2b-dev/e2b) to sandbox file and command operations
```
```suggestion
* We have an E2BRuntime, which uses e2b.dev containers to sandbox file and command operations
```

File: opendevin/README.md (hunk 2 above)
Comment (rbren) [reply opendevin/README.md:49]: Done!

File: opendevin/README.md [hunk 5]
```diff
@@ -1,25 +1,51 @@
...
-## Sandbox Image
-```bash
-docker build -f opendevin/sandbox/docker/Dockerfile -t opendevin/sandbox:v0.1 .
+# OpenDevin Architecture
+
+This directory contains the core components of OpenDevin.
+
+## Classes
+The key classes in OpenDevin are:
+* LLM: brokers all interactions with large language models. Works with any underlying completion model, thanks to LiteLLM.
```
Comment (rbren) [inline opendevin/README.md:7]:
```suggestion

* LLM: brokers all interactions with large language models. Works with any underlying completion model, thanks to LiteLLM.
```

**Comment by rbren** [comment]
Thanks all!

**Pull request merged**